### PR TITLE
[POS] Set the fiscal position on the shop and customer

### PR DESCRIPTION
**Pull request by ppa-odoo**
TASK : https://www.odoo.com/web?#id=11404&view_type=form&model=project.task&menu_id=3942&action=327
PAD : https://pad.odoo.com/p/r.onCx0SYmwpXjbJVJ

### CODE REVIEW

File: addons/point_of_sale/static/src/js/models.js [hunk 1]
```diff
@@ -1079,11 +1093,45 @@ exports.Orderline = Backbone.Model.extend({
...
+        var fiscal_id = this.get_fiscal_position_id();
+        var product = this.get_product().taxes_id;
+        var ptaxes_ids = [];
+        if(product && fiscal_id){
+            var fposition_tax = this.pos.fiscal_position_tax;
+            for(var f = 0; f < fposition_tax.length; f++ ){
+                if(fposition_tax[f].position_id[0] == fiscal_id){
+                    for(var p = 0; p < product.length; p++){
+                        ptaxes_ids[p] = (product[p] == fposition_tax[f].tax_src_id[0]) ? fposition_tax[f].tax_dest_id[0] : ptaxes_ids[p] || product[p];
+                    }
```
Comment (vme-odoo) [inline addons/point_of_sale/static/src/js/models.js]: ptaxes_ids[p] || product[p] in that, ptaxes_ids don't have value every time.

**Comment by vme-odoo** [comment]
Its better to store fiscal_position_id and its related taxes of source/destination tax id in customer/client widget, when set or change customer and then calculate applicable taxes from client taxes and product taxes.

**Comment by ppa-odoo** [comment]
@vme-odoo done as per your suggestions. 
Thanks

File: addons/point_of_sale/static/src/js/models.js [hunk 2]
```diff
@@ -1771,10 +1800,37 @@ exports.Order = Backbone.Model.extend({
...
+        var fposition = this.pos.fiscal_position;
+        if(client){
+            if(client.property_account_position[0]){
+                return client.property_account_position[0]
+            } else if (client.country_id){
+                for(var f = 0; f < fposition.length; f++){
+                    if(fposition[f].auto_apply == true && fposition[f].country_id[0] == client.country_id[0] ){
+                    return fposition[f].id;
+                    }
+                }
```
Comment (vme-odoo) [inline addons/point_of_sale/static/src/js/models.js]: Indent code, and use while loop, no need to check all cases.

File: addons/point_of_sale/static/src/js/models.js [hunk 3]
```diff
@@ -1771,10 +1800,37 @@ exports.Order = Backbone.Model.extend({
...
+            }
+        }
+        return false;
+    },
     /* ---- Client / Customer --- */
     // the client related to the current order.
     set_client: function(client){
         this.assert_editable();
+        this.fiscal_position_id = this.get_fiscal_position_id(client);
+        this.get_fiscal_tax(this.fiscal_position_id);
```
Comment (vme-odoo) [inline addons/point_of_sale/static/src/js/models.js]: rename function name its setter method, so set_fiscal_tax

File: addons/point_of_sale/static/src/js/models.js [hunk 4]
```diff
@@ -1771,10 +1800,37 @@ exports.Order = Backbone.Model.extend({
...
+    get_fiscal_tax: function(fiscal_id){
+        var fposition_tax = this.pos.fiscal_position_tax;
+        if(fiscal_id){
+            for(var f = 0; f < fposition_tax.length; f++ ){
+                if(fposition_tax[f].position_id[0] == fiscal_id){
+                    this.fiscal_tax[fposition_tax[f].tax_src_id[0]] = fposition_tax[f].tax_dest_id[0];
+                }
+            }
+        }
+    },
```
Comment (vme-odoo) [inline addons/point_of_sale/static/src/js/models.js]: improve code with return object and assign to this.fiscal_tax.

File: addons/point_of_sale/static/src/js/models.js [hunk 5]
```diff
@@ -1079,11 +1092,24 @@ exports.Orderline = Backbone.Model.extend({
...
+        var fiscal_tax = this.order.fiscal_tax;
+        var ptaxes_ids = [];
+        if(product_tax && fiscal_id){
+            for(var p = 0; p < product_tax.length; p++){
+                ptaxes_ids[p] = fiscal_tax.hasOwnProperty(product_tax[p]) ? fiscal_tax[product_tax[p]] : product_tax[p];
+            }
+            return ptaxes_ids;
+        }
+        return product_tax;
+    },
```
Comment (vme-odoo) [inline addons/point_of_sale/static/src/js/models.js]: remove  ptaxes_ids, assign direct into product_tax variable.

File: addons/point_of_sale/static/src/js/models.js (hunk 4 above)
Comment (ppa-odoo) [inline addons/point_of_sale/static/src/js/models.js]: done

File: addons/point_of_sale/static/src/js/models.js (hunk 2 above)
Comment (ppa-odoo) [inline addons/point_of_sale/static/src/js/models.js]: done

File: addons/point_of_sale/static/src/js/models.js (hunk 5 above)
Comment (ppa-odoo) [inline addons/point_of_sale/static/src/js/models.js]: @vme-odoo remove ptaxes_ids, 
Thanks

**Comment by hmo-odoo** [comment]
cancelled

**Pull request closed without merging**